// @ngx-spotify-api/core/esm2015/lib/ngx-spotify-api.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { AlbumService, ArtistService, BrowseService, FollowService, LibraryService, PersonalizationService, PlayerService, PlaylistService, SearchService, TrackService, UserService } from './services';
import { httpInterceptorProviders } from './http-interceptors';
import { ApiConfig, AuthorizationConfig } from './models/configs';
export class NgxSpotifyApiModule {
    /**
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        config = {
            api: Object.assign({}, new ApiConfig(), config.api),
            authorization: Object.assign({}, new AuthorizationConfig(), config.authorization)
        };
        return {
            ngModule: NgxSpotifyApiModule,
            providers: [
                { provide: AuthorizationConfig, useValue: config.authorization },
                { provide: ApiConfig, useValue: config.api },
            ]
        };
    }
}
NgxSpotifyApiModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule,
                ],
                providers: [
                    ArtistService,
                    AlbumService,
                    BrowseService,
                    FollowService,
                    LibraryService,
                    PersonalizationService,
                    PlayerService,
                    PlaylistService,
                    SearchService,
                    TrackService,
                    UserService,
                    httpInterceptorProviders
                ]
            },] }
];
