// @ngx-spotify-api/core/esm5/lib/services/artist.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { of } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
var ArtistListResponse = /** @class */ (function () {
    function ArtistListResponse() {
    }
    return ArtistListResponse;
}());
if (false) {
    /** @type {?} */
    ArtistListResponse.prototype.artists;
}
var TrackListResponse = /** @class */ (function () {
    function TrackListResponse() {
    }
    return TrackListResponse;
}());
if (false) {
    /** @type {?} */
    TrackListResponse.prototype.tracks;
}
var ArtistService = /** @class */ (function () {
    function ArtistService(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    ArtistService.prototype.getArtist = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id, {
            params: {}
        });
    };
    /**
     * @param {?} id
     * @param {?=} includeGroups
     * @param {?=} market
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    ArtistService.prototype.getArtistsAlbums = /**
     * @param {?} id
     * @param {?=} includeGroups
     * @param {?=} market
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    function (id, includeGroups, market, limit, offset) {
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id + '/albums', {
            params: {
                include_groups: includeGroups ? includeGroups.join(',') : null,
                market: market,
                limit: limit.toString(),
                offset: offset.toString()
            }
        });
    };
    /**
     * @param {?} id
     * @param {?} market
     * @return {?}
     */
    ArtistService.prototype.getArtistsTopTracks = /**
     * @param {?} id
     * @param {?} market
     * @return {?}
     */
    function (id, market) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id + '/top-tracks', {
            params: {
                market: market
            }
        }).pipe(map(function (res) {
            return res.tracks;
        }));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ArtistService.prototype.getArtistsRelatedArtists = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id + '/related-artists', {
            params: {}
        }).pipe(map(function (res) {
            return res.artists;
        }));
    };
    /**
     * @param {?} ids
     * @return {?}
     */
    ArtistService.prototype.getArtists = /**
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        if (ids == null || ids.length < 1) {
            return of([]);
        }
        else if (ids.length > 50) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + '/v1/artists/', {
            params: {
                ids: ids.join(',')
            }
        }).pipe(map(function (res) {
            return res.artists;
        }));
    };
    ArtistService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ArtistService.ctorParameters = function () { return [
        { type: ApiConfig, decorators: [{ type: Optional }] },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ ArtistService.ngInjectableDef = i0.defineInjectable({ factory: function ArtistService_Factory() { return new ArtistService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: ArtistService, providedIn: "root" });
    return ArtistService;
}());
export { ArtistService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ArtistService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    ArtistService.prototype.http;
}
