// @ngx-spotify-api/core/esm5/lib/services/browse.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
var PlaylistList = /** @class */ (function () {
    function PlaylistList() {
    }
    return PlaylistList;
}());
if (false) {
    /** @type {?} */
    PlaylistList.prototype.playlists;
}
var CategoryList = /** @class */ (function () {
    function CategoryList() {
    }
    return CategoryList;
}());
if (false) {
    /** @type {?} */
    CategoryList.prototype.categories;
}
var BrowseService = /** @class */ (function () {
    function BrowseService(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @param {?=} country
     * @param {?=} locale
     * @return {?}
     */
    BrowseService.prototype.getCategory = /**
     * @param {?} id
     * @param {?=} country
     * @param {?=} locale
     * @return {?}
     */
    function (id, country, locale) {
        return this.http.get(this.config.baseUrl + '/v1/browse/categories/' + id, {
            params: {
                country: country,
                locale: locale
            }
        });
    };
    /**
     * @param {?} id
     * @param {?=} country
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    BrowseService.prototype.getCategoryPlaylists = /**
     * @param {?} id
     * @param {?=} country
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    function (id, country, limit, offset) {
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + '/v1/browse/categories/' + id + '/playlists', {
            params: {
                country: country,
                limit: limit.toString(),
                offset: offset.toString()
            }
        }).pipe(map(function (res) {
            return res.playlists;
        }));
    };
    /**
     * @param {?=} country
     * @param {?=} locale
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    BrowseService.prototype.getCategories = /**
     * @param {?=} country
     * @param {?=} locale
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    function (country, locale, limit, offset) {
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + '/v1/browse/categories', {
            params: {
                country: country,
                locale: locale,
                limit: limit.toString(),
                offset: offset.toString()
            }
        }).pipe(map(function (list) {
            return list.categories;
        }));
    };
    /**
     * @param {?=} country
     * @param {?=} locale
     * @param {?=} timestamp
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    BrowseService.prototype.getFeaturedPlaylists = /**
     * @param {?=} country
     * @param {?=} locale
     * @param {?=} timestamp
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    function (country, locale, timestamp, limit, offset) {
        if (timestamp === void 0) { timestamp = new Date(); }
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + '/v1/browse/featured-playlists', {
            params: {
                country: country,
                locale: locale,
                timestamp: timestamp.toISOString(),
                limit: limit.toString(),
                offset: offset.toString()
            }
        });
    };
    /**
     * @param {?=} country
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    BrowseService.prototype.getNewReleases = /**
     * @param {?=} country
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    function (country, limit, offset) {
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + '/v1/browse/new-releases', {
            params: {
                country: country,
                limit: limit.toString(),
                offset: offset.toString()
            }
        }).pipe(map(function (_a) {
            var albums = _a.albums;
            return albums;
        }));
    };
    BrowseService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    BrowseService.ctorParameters = function () { return [
        { type: ApiConfig, decorators: [{ type: Optional }] },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ BrowseService.ngInjectableDef = i0.defineInjectable({ factory: function BrowseService_Factory() { return new BrowseService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: BrowseService, providedIn: "root" });
    return BrowseService;
}());
export { BrowseService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BrowseService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    BrowseService.prototype.http;
}
