// @ngx-spotify-api/core/esm5/lib/services/personalization.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
var PersonalizationService = /** @class */ (function () {
    function PersonalizationService(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?=} limit
     * @param {?=} offset
     * @param {?=} timeRange
     * @return {?}
     */
    PersonalizationService.prototype.getTopArtists = /**
     * @param {?=} limit
     * @param {?=} offset
     * @param {?=} timeRange
     * @return {?}
     */
    function (limit, offset, timeRange) {
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + "/v1/me/top/artists", {
            params: {
                limit: limit.toString(),
                offset: offset.toString(),
                time_range: timeRange
            }
        });
    };
    /**
     * @param {?=} limit
     * @param {?=} offset
     * @param {?=} timeRange
     * @return {?}
     */
    PersonalizationService.prototype.getTopTracks = /**
     * @param {?=} limit
     * @param {?=} offset
     * @param {?=} timeRange
     * @return {?}
     */
    function (limit, offset, timeRange) {
        if (limit === void 0) { limit = 20; }
        if (offset === void 0) { offset = 0; }
        return this.http.get(this.config.baseUrl + "/v1/me/top/tracks", {
            params: {
                limit: limit.toString(),
                offset: offset.toString(),
                time_range: timeRange
            }
        });
    };
    PersonalizationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PersonalizationService.ctorParameters = function () { return [
        { type: ApiConfig, decorators: [{ type: Optional }] },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ PersonalizationService.ngInjectableDef = i0.defineInjectable({ factory: function PersonalizationService_Factory() { return new PersonalizationService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: PersonalizationService, providedIn: "root" });
    return PersonalizationService;
}());
export { PersonalizationService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PersonalizationService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    PersonalizationService.prototype.http;
}
