// @ngx-spotify-api/core/esm5/lib/services/track.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
var TrackService = /** @class */ (function () {
    function TrackService(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @param {?=} market
     * @return {?}
     */
    TrackService.prototype.getTrack = /**
     * @param {?} id
     * @param {?=} market
     * @return {?}
     */
    function (id, market) {
        return this.http.get(this.config.baseUrl + ("/v1/tracks/" + id), {
            params: {
                market: market
            }
        });
    };
    /**
     * @param {?} id
     * @return {?}
     */
    TrackService.prototype.getAudioFeatures = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get(this.config.baseUrl + ("/v1/audio-features/" + id));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    TrackService.prototype.getAudioAnalysis = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.http.get(this.config.baseUrl + ("/v1/audio-analysis/" + id));
    };
    /**
     * @param {?} ids
     * @param {?=} market
     * @return {?}
     */
    TrackService.prototype.getTracks = /**
     * @param {?} ids
     * @param {?=} market
     * @return {?}
     */
    function (ids, market) {
        if (ids === undefined || ids === null || ids.length === 0) {
            return of([]);
        }
        else if (ids.length > 50) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + "/v1/tracks", {
            params: {
                ids: ids.join(','),
                market: market || ''
            }
        }).pipe(map(function (_a) {
            var tracks = _a.tracks;
            return tracks;
        }));
    };
    /**
     * @param {?} ids
     * @return {?}
     */
    TrackService.prototype.getSeveralAudioFeatures = /**
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        if (ids === undefined || ids === null || ids.length === 0) {
            return of([]);
        }
        else if (ids.length > 50) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + "/v1/audio-features", {
            params: {
                ids: ids.join(',')
            }
        }).pipe(map(function (_a) {
            var audioFeatures = _a.audioFeatures;
            return audioFeatures;
        }));
    };
    TrackService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TrackService.ctorParameters = function () { return [
        { type: ApiConfig, decorators: [{ type: Optional }] },
        { type: HttpClient }
    ]; };
    /** @nocollapse */ TrackService.ngInjectableDef = i0.defineInjectable({ factory: function TrackService_Factory() { return new TrackService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: TrackService, providedIn: "root" });
    return TrackService;
}());
export { TrackService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TrackService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    TrackService.prototype.http;
}
