// @ngx-spotify-api/core/esm5/lib/http-interceptors/Auth.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ApiConfig, AuthorizationConfig } from '../models/configs';
import { flatMap, map } from 'rxjs/operators';
var AuthInterceptor = /** @class */ (function () {
    function AuthInterceptor(apiConfig, authConfig) {
        this.apiConfig = apiConfig;
        this.authConfig = authConfig;
    }
    /**
     * @private
     * @param {?} req
     * @return {?}
     */
    AuthInterceptor.prototype.attachAuthorizationHeader = /**
     * @private
     * @param {?} req
     * @return {?}
     */
    function (req) {
        return this.authConfig.accessToken().pipe(map(function (token) {
            return req.clone({
                headers: req.headers.set('Authorization', 'Bearer ' + token)
            });
        }));
    };
    /**
     * @param {?} origReq
     * @param {?} next
     * @return {?}
     */
    AuthInterceptor.prototype.intercept = /**
     * @param {?} origReq
     * @param {?} next
     * @return {?}
     */
    function (origReq, next) {
        return this.attachAuthorizationHeader(origReq).pipe(flatMap(function (req) {
            return next.handle(req);
        }));
    };
    AuthInterceptor.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuthInterceptor.ctorParameters = function () { return [
        { type: ApiConfig },
        { type: AuthorizationConfig }
    ]; };
    return AuthInterceptor;
}());
export { AuthInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthInterceptor.prototype.apiConfig;
    /**
     * @type {?}
     * @private
     */
    AuthInterceptor.prototype.authConfig;
}
