// @ngx-spotify-api/core/esm5/lib/http-interceptors/UnderscoreToCamelcase.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { Injectable } from '@angular/core';
import { ApiConfig } from '../models/configs/api.config';
import { AuthorizationConfig } from "../models/configs";
var UnderscoreToCamelcaseInterceptor = /** @class */ (function () {
    function UnderscoreToCamelcaseInterceptor(apiConfig, authConfig) {
        this.apiConfig = apiConfig;
        this.authConfig = authConfig;
    }
    /**
     * @private
     * @param {?} object
     * @return {?}
     */
    UnderscoreToCamelcaseInterceptor.prototype.transformObjectKeysToCamelCase = /**
     * @private
     * @param {?} object
     * @return {?}
     */
    function (object) {
        var _this = this;
        if (object == null || object === {}) {
            return object;
        }
        /** @type {?} */
        var ret = {};
        Object.keys(object).forEach(function (key) {
            /** @type {?} */
            var newKey = key.replace(/[_]([^_])?/g, function (v) {
                return v.replace('_', '').toUpperCase();
            });
            if (object[key] instanceof Array) {
                ret[newKey] = object[key].map(function (value) {
                    return _this.transformObjectKeysToCamelCase(value);
                });
            }
            else if (object[key] instanceof Object) {
                ret[newKey] = _this.transformObjectKeysToCamelCase(object[key]);
            }
            else {
                ret[newKey] = object[key];
            }
        });
        return ret;
    };
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    UnderscoreToCamelcaseInterceptor.prototype.intercept = /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        return next.handle(req).pipe(map(function (res) {
            if (req && req.url && req.url.includes(_this.apiConfig.baseUrl) && res.body) {
                return res.clone({
                    body: _this.transformObjectKeysToCamelCase(res.body)
                });
            }
            return res;
        }));
    };
    UnderscoreToCamelcaseInterceptor.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UnderscoreToCamelcaseInterceptor.ctorParameters = function () { return [
        { type: ApiConfig },
        { type: AuthorizationConfig }
    ]; };
    return UnderscoreToCamelcaseInterceptor;
}());
export { UnderscoreToCamelcaseInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UnderscoreToCamelcaseInterceptor.prototype.apiConfig;
    /**
     * @type {?}
     * @private
     */
    UnderscoreToCamelcaseInterceptor.prototype.authConfig;
}
