// @ngx-spotify-api/core/esm2015/lib/http-interceptors/Auth.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ApiConfig, AuthorizationConfig } from '../models/configs';
import { flatMap, map } from 'rxjs/operators';
export class AuthInterceptor {
    /**
     * @param {?} apiConfig
     * @param {?} authConfig
     */
    constructor(apiConfig, authConfig) {
        this.apiConfig = apiConfig;
        this.authConfig = authConfig;
    }
    /**
     * @private
     * @param {?} req
     * @return {?}
     */
    attachAuthorizationHeader(req) {
        return this.authConfig.accessToken().pipe(map((token) => {
            return req.clone({
                headers: req.headers.set('Authorization', 'Bearer ' + token)
            });
        }));
    }
    /**
     * @param {?} origReq
     * @param {?} next
     * @return {?}
     */
    intercept(origReq, next) {
        return this.attachAuthorizationHeader(origReq).pipe(flatMap((req) => {
            return next.handle(req);
        }));
    }
}
AuthInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthInterceptor.ctorParameters = () => [
    { type: ApiConfig },
    { type: AuthorizationConfig }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthInterceptor.prototype.apiConfig;
    /**
     * @type {?}
     * @private
     */
    AuthInterceptor.prototype.authConfig;
}
