// @ngx-spotify-api/core/esm2015/lib/http-interceptors/UnderscoreToCamelcase.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { Injectable } from '@angular/core';
import { ApiConfig } from '../models/configs/api.config';
import { AuthorizationConfig } from "../models/configs";
export class UnderscoreToCamelcaseInterceptor {
    /**
     * @param {?} apiConfig
     * @param {?} authConfig
     */
    constructor(apiConfig, authConfig) {
        this.apiConfig = apiConfig;
        this.authConfig = authConfig;
    }
    /**
     * @private
     * @param {?} object
     * @return {?}
     */
    transformObjectKeysToCamelCase(object) {
        if (object == null || object === {}) {
            return object;
        }
        /** @type {?} */
        const ret = {};
        Object.keys(object).forEach((key) => {
            /** @type {?} */
            const newKey = key.replace(/[_]([^_])?/g, (v) => {
                return v.replace('_', '').toUpperCase();
            });
            if (object[key] instanceof Array) {
                ret[newKey] = object[key].map((value) => {
                    return this.transformObjectKeysToCamelCase(value);
                });
            }
            else if (object[key] instanceof Object) {
                ret[newKey] = this.transformObjectKeysToCamelCase(object[key]);
            }
            else {
                ret[newKey] = object[key];
            }
        });
        return ret;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        return next.handle(req).pipe(map((res) => {
            if (req && req.url && req.url.includes(this.apiConfig.baseUrl) && res.body) {
                return res.clone({
                    body: this.transformObjectKeysToCamelCase(res.body)
                });
            }
            return res;
        }));
    }
}
UnderscoreToCamelcaseInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
UnderscoreToCamelcaseInterceptor.ctorParameters = () => [
    { type: ApiConfig },
    { type: AuthorizationConfig }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    UnderscoreToCamelcaseInterceptor.prototype.apiConfig;
    /**
     * @type {?}
     * @private
     */
    UnderscoreToCamelcaseInterceptor.prototype.authConfig;
}
