// @ngx-spotify-api/core/esm5/lib/ngx-spotify-api.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { AlbumService, ArtistService, BrowseService, FollowService, LibraryService, PersonalizationService, PlayerService, PlaylistService, SearchService, TrackService, UserService } from './services';
import { httpInterceptorProviders } from './http-interceptors';
import { ApiConfig, AuthorizationConfig } from './models/configs';
var NgxSpotifyApiModule = /** @class */ (function () {
    function NgxSpotifyApiModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    NgxSpotifyApiModule.forRoot = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        config = {
            api: tslib_1.__assign({}, new ApiConfig(), config.api),
            authorization: tslib_1.__assign({}, new AuthorizationConfig(), config.authorization)
        };
        return {
            ngModule: NgxSpotifyApiModule,
            providers: [
                { provide: AuthorizationConfig, useValue: config.authorization },
                { provide: ApiConfig, useValue: config.api },
            ]
        };
    };
    NgxSpotifyApiModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                    ],
                    providers: [
                        ArtistService,
                        AlbumService,
                        BrowseService,
                        FollowService,
                        LibraryService,
                        PersonalizationService,
                        PlayerService,
                        PlaylistService,
                        SearchService,
                        TrackService,
                        UserService,
                        httpInterceptorProviders
                    ]
                },] }
    ];
    return NgxSpotifyApiModule;
}());
export { NgxSpotifyApiModule };
