// @ngx-spotify-api/core/esm2015/lib/services/album.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
class AlbumsResponse {
}
if (false) {
    /** @type {?} */
    AlbumsResponse.prototype.albums;
}
export class AlbumService {
    /**
     * @param {?} config
     * @param {?} http
     */
    constructor(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @param {?=} market
     * @return {?}
     */
    getAlbum(id, market) {
        return this.http.get(this.config.baseUrl + '/v1/albums/' + id, {
            params: {
                market: market
            }
        });
    }
    /**
     * @param {?} id
     * @param {?=} limit
     * @param {?=} offset
     * @param {?=} market
     * @return {?}
     */
    getAlbumTracks(id, limit = 20, offset = 0, market) {
        return this.http.get(this.config.baseUrl + '/v1/albums/' + id + '/tracks', {
            params: {
                limit: limit.toString(),
                offset: offset.toString(),
                market: market
            }
        });
    }
    /**
     * @param {?} ids
     * @param {?=} market
     * @return {?}
     */
    getAlbums(ids, market) {
        if (ids == null || ids.length < 1) {
            return of([]);
        }
        else if (ids.length > 20) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + '/v1/albums', {
            params: {
                ids: ids.join(','),
                market: market
            }
        }).pipe(map((res) => {
            return res.albums;
        }));
    }
}
AlbumService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AlbumService.ctorParameters = () => [
    { type: ApiConfig, decorators: [{ type: Optional }] },
    { type: HttpClient }
];
/** @nocollapse */ AlbumService.ngInjectableDef = i0.defineInjectable({ factory: function AlbumService_Factory() { return new AlbumService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: AlbumService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AlbumService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    AlbumService.prototype.http;
}
