// @ngx-spotify-api/core/esm2015/lib/services/artist.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { of } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
class ArtistListResponse {
}
if (false) {
    /** @type {?} */
    ArtistListResponse.prototype.artists;
}
class TrackListResponse {
}
if (false) {
    /** @type {?} */
    TrackListResponse.prototype.tracks;
}
export class ArtistService {
    /**
     * @param {?} config
     * @param {?} http
     */
    constructor(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getArtist(id) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id, {
            params: {}
        });
    }
    /**
     * @param {?} id
     * @param {?=} includeGroups
     * @param {?=} market
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    getArtistsAlbums(id, includeGroups, market, limit = 20, offset = 0) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id + '/albums', {
            params: {
                include_groups: includeGroups ? includeGroups.join(',') : null,
                market,
                limit: limit.toString(),
                offset: offset.toString()
            }
        });
    }
    /**
     * @param {?} id
     * @param {?} market
     * @return {?}
     */
    getArtistsTopTracks(id, market) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id + '/top-tracks', {
            params: {
                market: market
            }
        }).pipe(map((res) => {
            return res.tracks;
        }));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getArtistsRelatedArtists(id) {
        return this.http.get(this.config.baseUrl + '/v1/artists/' + id + '/related-artists', {
            params: {}
        }).pipe(map((res) => {
            return res.artists;
        }));
    }
    /**
     * @param {?} ids
     * @return {?}
     */
    getArtists(ids) {
        if (ids == null || ids.length < 1) {
            return of([]);
        }
        else if (ids.length > 50) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + '/v1/artists/', {
            params: {
                ids: ids.join(',')
            }
        }).pipe(map((res) => {
            return res.artists;
        }));
    }
}
ArtistService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ArtistService.ctorParameters = () => [
    { type: ApiConfig, decorators: [{ type: Optional }] },
    { type: HttpClient }
];
/** @nocollapse */ ArtistService.ngInjectableDef = i0.defineInjectable({ factory: function ArtistService_Factory() { return new ArtistService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: ArtistService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ArtistService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    ArtistService.prototype.http;
}
