// @ngx-spotify-api/core/esm2015/lib/services/browse.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
class PlaylistList {
}
if (false) {
    /** @type {?} */
    PlaylistList.prototype.playlists;
}
class CategoryList {
}
if (false) {
    /** @type {?} */
    CategoryList.prototype.categories;
}
export class BrowseService {
    /**
     * @param {?} config
     * @param {?} http
     */
    constructor(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @param {?=} country
     * @param {?=} locale
     * @return {?}
     */
    getCategory(id, country, locale) {
        return this.http.get(this.config.baseUrl + '/v1/browse/categories/' + id, {
            params: {
                country: country,
                locale: locale
            }
        });
    }
    /**
     * @param {?} id
     * @param {?=} country
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    getCategoryPlaylists(id, country, limit = 20, offset = 0) {
        return this.http.get(this.config.baseUrl + '/v1/browse/categories/' + id + '/playlists', {
            params: {
                country: country,
                limit: limit.toString(),
                offset: offset.toString()
            }
        }).pipe(map((res) => {
            return res.playlists;
        }));
    }
    /**
     * @param {?=} country
     * @param {?=} locale
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    getCategories(country, locale, limit = 20, offset = 0) {
        return this.http.get(this.config.baseUrl + '/v1/browse/categories', {
            params: {
                country: country,
                locale: locale,
                limit: limit.toString(),
                offset: offset.toString()
            }
        }).pipe(map((list) => {
            return list.categories;
        }));
    }
    /**
     * @param {?=} country
     * @param {?=} locale
     * @param {?=} timestamp
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    getFeaturedPlaylists(country, locale, timestamp = new Date(), limit = 20, offset = 0) {
        return this.http.get(this.config.baseUrl + '/v1/browse/featured-playlists', {
            params: {
                country,
                locale,
                timestamp: timestamp.toISOString(),
                limit: limit.toString(),
                offset: offset.toString()
            }
        });
    }
    /**
     * @param {?=} country
     * @param {?=} limit
     * @param {?=} offset
     * @return {?}
     */
    getNewReleases(country, limit = 20, offset = 0) {
        return this.http.get(this.config.baseUrl + '/v1/browse/new-releases', {
            params: {
                country,
                limit: limit.toString(),
                offset: offset.toString()
            }
        }).pipe(map(({ albums }) => {
            return albums;
        }));
    }
}
BrowseService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
BrowseService.ctorParameters = () => [
    { type: ApiConfig, decorators: [{ type: Optional }] },
    { type: HttpClient }
];
/** @nocollapse */ BrowseService.ngInjectableDef = i0.defineInjectable({ factory: function BrowseService_Factory() { return new BrowseService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: BrowseService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    BrowseService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    BrowseService.prototype.http;
}
