// @ngx-spotify-api/core/esm2015/lib/services/player.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map, take } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
class DeviceList {
}
if (false) {
    /** @type {?} */
    DeviceList.prototype.devices;
}
export class PlayerService {
    /**
     * @param {?} config
     * @param {?} http
     */
    constructor(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?=} deviceId
     * @return {?}
     */
    nextTrack(deviceId) {
        return this.http.post(this.config.baseUrl + `/v1/me/player/next`, {}, {
            params: {
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @param {?=} deviceId
     * @return {?}
     */
    previousTrack(deviceId) {
        return this.http.post(this.config.baseUrl + `/v1/me/player/previous`, {}, {
            params: {
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @param {?} positionMs
     * @param {?=} deviceId
     * @return {?}
     */
    seekToPosition(positionMs, deviceId) {
        return this.http.put(this.config.baseUrl + `/v1/me/player/seek`, {}, {
            params: {
                position_ms: positionMs.toString(),
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @return {?}
     */
    getAvailableDevices() {
        return this.http.get(this.config.baseUrl + `/v1/me/player/devices`).pipe(map((list) => list.devices));
    }
    /**
     * @param {?} shuffle
     * @param {?=} deviceId
     * @return {?}
     */
    toggleShuffle(shuffle, deviceId) {
        return this.http.put(this.config.baseUrl + `/v1/me/player/shuffle`, {}, {
            params: {
                state: shuffle.toString(),
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     *
     * @param {?} deviceIds Although an array is accepted, only a single deviceId is currently supported
     * @param {?=} forcePlay
     * @return {?}
     */
    transferPlayback(deviceIds, forcePlay = false) {
        return this.http.put(this.config.baseUrl + `/v1/me/player`, {
            device_ids: deviceIds.join(','),
            play: forcePlay
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @param {?=} limit
     * @param {?=} before
     * @param {?=} after
     * @return {?}
     */
    getRecentlyPlayedTracks(limit = 20, before, after = 0) {
        return this.http.get(this.config.baseUrl + `/v1/me/player/recently-played`, {
            params: {
                limit: limit.toString(),
                before: (before ? before.toString() : undefined),
                after: after.toString(),
            }
        });
    }
    /**
     * @param {?=} deviceId
     * @param {?=} context
     * @param {?=} uris
     * @param {?=} offset
     * @param {?=} positionMs
     * @return {?}
     */
    startPlayback(deviceId, context, uris, offset, positionMs) {
        return this.http.put(this.config.baseUrl + `/v1/me/player/play`, {
            context_uri: context,
            uris: uris,
            offset: offset,
            position_ms: positionMs
        }, {
            params: {
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @param {?=} deviceId
     * @return {?}
     */
    pausePlayback(deviceId) {
        return this.http.put(this.config.baseUrl + `/v1/me/player/pause`, {}, {
            params: {
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @param {?} mode
     * @param {?=} deviceId
     * @return {?}
     */
    toggleRepeatMode(mode, deviceId) {
        return this.http.put(this.config.baseUrl + `/v1/me/player/repeat`, {}, {
            params: {
                state: mode,
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
    /**
     * @param {?=} market
     * @return {?}
     */
    getPlayback(market) {
        return this.http.get(this.config.baseUrl + `/v1/me/player`, {
            params: {
                market: market
            }
        });
    }
    /**
     * @param {?=} market
     * @return {?}
     */
    getCurrentlyPlayingTrack(market) {
        return this.http.get(this.config.baseUrl + `/v1/me/player/currently-playing`, {
            params: {
                market: market
            }
        });
    }
    /**
     * @param {?} volume
     * @param {?=} deviceId
     * @return {?}
     */
    setPlaybackVolume(volume, deviceId) {
        return this.http.put(this.config.baseUrl + `/v1/me/player/volume`, {}, {
            params: {
                volume_percent: volume.toString(),
                device_id: deviceId
            }
        }).pipe(take(1), map(() => {
            return true;
        }));
    }
}
PlayerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PlayerService.ctorParameters = () => [
    { type: ApiConfig, decorators: [{ type: Optional }] },
    { type: HttpClient }
];
/** @nocollapse */ PlayerService.ngInjectableDef = i0.defineInjectable({ factory: function PlayerService_Factory() { return new PlayerService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: PlayerService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PlayerService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    PlayerService.prototype.http;
}
