// @ngx-spotify-api/core/esm2015/lib/services/track.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiConfig } from '../models';
import * as i0 from "@angular/core";
import * as i1 from "../models/configs/api.config";
import * as i2 from "@angular/common/http";
export class TrackService {
    /**
     * @param {?} config
     * @param {?} http
     */
    constructor(config, http) {
        this.config = config;
        this.http = http;
    }
    /**
     * @param {?} id
     * @param {?=} market
     * @return {?}
     */
    getTrack(id, market) {
        return this.http.get(this.config.baseUrl + `/v1/tracks/${id}`, {
            params: {
                market
            }
        });
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getAudioFeatures(id) {
        return this.http.get(this.config.baseUrl + `/v1/audio-features/${id}`);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getAudioAnalysis(id) {
        return this.http.get(this.config.baseUrl + `/v1/audio-analysis/${id}`);
    }
    /**
     * @param {?} ids
     * @param {?=} market
     * @return {?}
     */
    getTracks(ids, market) {
        if (ids === undefined || ids === null || ids.length === 0) {
            return of([]);
        }
        else if (ids.length > 50) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + `/v1/tracks`, {
            params: {
                ids: ids.join(','),
                market: market || ''
            }
        }).pipe(map(({ tracks }) => tracks));
    }
    /**
     * @param {?} ids
     * @return {?}
     */
    getSeveralAudioFeatures(ids) {
        if (ids === undefined || ids === null || ids.length === 0) {
            return of([]);
        }
        else if (ids.length > 50) {
            return of(null);
        }
        return this.http.get(this.config.baseUrl + `/v1/audio-features`, {
            params: {
                ids: ids.join(',')
            }
        }).pipe(map(({ audioFeatures }) => audioFeatures));
    }
}
TrackService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TrackService.ctorParameters = () => [
    { type: ApiConfig, decorators: [{ type: Optional }] },
    { type: HttpClient }
];
/** @nocollapse */ TrackService.ngInjectableDef = i0.defineInjectable({ factory: function TrackService_Factory() { return new TrackService(i0.inject(i1.ApiConfig, 8), i0.inject(i2.HttpClient)); }, token: TrackService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TrackService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    TrackService.prototype.http;
}
